### ISSUE TITLE
Memory Leak of ModuleBuilder

### ISSUE THREAD

**Issue by cyberden** [jehugaleahsa/FlatFiles#65]
Hello,

We are using FlatFile in a software used to write a flat file. 

We have a memory leak problem, The instances of ModuleBuilder are getting huge because each time we want to write the file, the DefineType method is called on the static ModuleBuilder with a new type name (because of the GetUniqueTypeName method). All this is in CodeGenerator.cs .

We don't understand why each time we go through this code a new type is defined whereas the existing type could be reused.

Is this a bug? Is our usage wrong?

Thanks for your time!

**Comment by Rayzbam** [comment]
Same issue here.

I have a dump file analysis of this bug, never found a solution, (disable OptimizeMapping seems not to be a good one) : 

![image](https://user-images.githubusercontent.com/52962035/93469288-942c6b80-f8f0-11ea-962f-c314adc49c3f.png)

![image](https://user-images.githubusercontent.com/52962035/93469357-b1f9d080-f8f0-11ea-8287-ca7c8f6dbae4.png)

![image](https://user-images.githubusercontent.com/52962035/93469380-b920de80-f8f0-11ea-8710-c99503bf340d.png)

![image](https://user-images.githubusercontent.com/52962035/93469411-c211b000-f8f0-11ea-8ee3-5a8f3304c8cc.png)

**Comment by jehugaleahsa** [comment]
I had to re-read this code again myself. I see the issue and will try to put out a fix soon - it should improve performance slightly in your case, too.

**Comment by jehugaleahsa** [comment]
I just put out a NuGet package that should hopefully resolve your issue: https://www.nuget.org/packages/FlatFiles/4.8.0

You can look at the changes here: https://github.com/jehugaleahsa/FlatFiles/commit/41566a2e0af65619a8212b0142e6c70abfd58a59

tl;dr I stopped storing the dynamic assembly in a `static` variable and now the assembly lives for as long as the `ICodeGenerator`. I had to update the assembly type to `RunAndCollect` in order to allow garbage collection to clean up after a while.

When the NuGet package becomes available (should be soon), I would ask that you do your best to try to see if the issue is resolved (I know diagnosing and reproducing this sort of issue is difficult). If everything looks good, would you please close this ticket?

Thanks for finding this! And for using FlatFiles!

**Comment by Rayzbam** [comment]
I'm glad to read this !

@jehugaleahsa Trust me, we're gonna try it very soon.

Thanks for your quick update

**Comment by Rayzbam** [comment]
@jehugaleahsa  We just tried on our side and it's much better.
Your package is pretty well designed and we're glad to found it for our projects.
Thanks for your update, we had this issue on our Production server and you helped us very quickly !

**Comment by cyberden** [comment]
Thanks for your fix. It will be deployed on our production env today and should hopefully fix our issues.

I close the issue. Thanks again for looking into the issue and the package :)
